llvm/tools/plugins-shlib/pypass.cpp

```diff
@@ -0,0 +1,501 @@
+
+  // dictobject.h
+  PyDict_GetItemString_t *PyDict_GetItemString;
+  PyDict_SetItemString_t *PyDict_SetItemString;
+  PyDict_New_t *PyDict_New;
+
+  // abstract.h
+  PyObject_CallObject_t *PyObject_CallObject;
+
+  // longobject.h
+  PyLong_FromVoidPtr_t *PyLong_FromVoidPtr;
+
+  // unicodeobject.h
+  PyUnicode_FromString_t *PyUnicode_FromString;
+
+  // tupleobject.h
+  PyTuple_SetItem_t *PyTuple_SetItem;
+  PyTuple_New_t *PyTuple_New;
+
+  void *PyGlobals;
+  void *PyBuiltins;
+
+private:
+  PythonAPI() : Valid(false) {
+    if (!loadDylib(findPython()))
+      return;
+    if (!resolveSymbols())
+      return;
+    Py_InitializeEx(0);
+    PyBuiltins = PyImport_ImportModule("builtins");
+    PyGlobals = PyModule_GetDict(PyImport_AddModule("__main__"));
+    Valid = true;
+  }
+
+  ~PythonAPI() {
+    if (std::atomic_exchange(&Valid, false)) {
+      Py_DecRef(PyBuiltins);
+      Py_DecRef(PyGlobals);
+      Py_FinalizeEx();
+    }
+  }
+
+  bool loadDylib(std::string Path) {
+    std::string Err;
+    Dylib = sys::DynamicLibrary::getPermanentLibrary(Path.c_str(), &Err);
+    if (!Dylib.isValid()) {
+      errs() << "dlopen for '" << Path << "' failed: " << Err << "\n";
+      return false;
+    }
+
+    return true;
+  }
+
+  bool resolveSymbols() {
+    bool Success = true;
+    Success &= resolve("_Py_IncRef", &Py_IncRef);
+    Success &= resolve("_Py_DecRef", &Py_DecRef);
+    Success &= resolve("Py_InitializeEx", &Py_InitializeEx);
+    Success &= resolve("Py_FinalizeEx", &Py_FinalizeEx);
+    Success &= resolve("PyErr_Print", &PyErr_Print);
+    Success &= resolve("PyGILState_Ensure", &PyGILState_Ensure);
+    Success &= resolve("PyGILState_Release", &PyGILState_Release);
+    Success &= resolve("PyImport_AddModule", &PyImport_AddModule);
+    Success &= resolve("PyImport_ImportModule", &PyImport_ImportModule);
+    Success &= resolve("PyModule_GetDict", &PyModule_GetDict);
+    Success &= resolve("PyDict_GetItemString", &PyDict_GetItemString);
+    Success &= resolve("PyDict_SetItemString", &PyDict_SetItemString);
+    Success &= resolve("PyDict_New", &PyDict_New);
+    Success &= resolve("PyObject_CallObject", &PyObject_CallObject);
+    Success &= resolve("PyObject_GetAttrString", &PyObject_GetAttrString);
+    Success &= resolve("PyObject_IsTrue", &PyObject_IsTrue);
+    Success &= resolve("PyLong_FromVoidPtr", &PyLong_FromVoidPtr);
+    Success &= resolve("PyUnicode_FromString", &PyUnicode_FromString);
+    Success &= resolve("PyTuple_SetItem", &PyTuple_SetItem);
+    Success &= resolve("PyTuple_New", &PyTuple_New);
+    return Success;
+  }
+
+  bool importModule(const char *Name) const {
+    void *Mod = PyImport_ImportModule(Name);
+    if (!Mod) {
+      PyErr_Print();
+      return false;
+    }
+    PyDict_SetItemString(PyGlobals, Name, Mod);
+    Py_DecRef(Mod);
+    return true;
+  }
+
+  bool evaluate(std::string Code, void *Globals) const {
+    void *Exec = PyObject_GetAttrString(PyBuiltins, "exec");
```

Comment:
`Exec` is a new Reference, you fail to release it in all cases there (e.g. when returning `false` below.

Reply:
It might be better to have RAII class helpers to handle the GIL and to release object references (thus also getting rid of the various make_scope_exit()).